File: iosApp/iosApp/Resources/Typography.swift
import iosExport
import SwiftUI

struct Typography {
    var text: TitleTypography { TitleTypography() }
    var main: MainTypography { MainTypography() }
    var menu: MenuTypography { MenuTypography() }
    var navigation: NavigationTypography { NavigationTypography() }
}

struct TitleTypography {
    var title: Font { .bold(size: 22) }
    var main: Font { .regular(size: 16) }
    var buttonText: Font { .bold(size: 18) }
    var inputText: Font { .regular(size: 16) }
    var hintText: Font { .regular(size: 16) }
}

struct MainTypography {
    var bigText: Font { .bold(size: 40) }
    var mediumText: Font { .bold(size: 32) }
    var text: Font { .bold(size: 24) }
    var button: Font { .regular(size: 20) }
}

struct MenuTypography {
    var title: Font { .regular(size: 24) }
    var tableDescription: Font { .regular(size: 16) }
    var itemKey: Font { .bold(size: 24) }
    var itemValue: Font { .regular(size: 24) }
}

struct NavigationTypography {
    var title: Font { .bold(size: 28) }
    var button: Font { .regular(size: 20) }
}

File: androidApp/ui/src/main/java/com/studcenter/ui/Typography.kt
package com.studcenter.ui


import androidx.compose.runtime.Immutable
import androidx.compose.ui.text.PlatformTextStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.LineHeightStyle
import androidx.compose.ui.unit.sp

@Immutable
data class Typography(private val regular: FontFamily, private val bold: FontFamily) {
    val text: TitleTypography = TitleTypography(regular, bold)
    val menu: MenuTypography = MenuTypography(regular, bold)
}

@Immutable
data class TitleTypography(
    private val regular: FontFamily,
    private val bold: FontFamily
) {
    val title: TextStyle = TextStyle(
        fontSize = 32.sp,
        lineHeight = 28.sp,
        fontFamily = bold,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()

    val main: TextStyle = TextStyle(
        fontSize = 16.sp,
        lineHeight = 28.sp,
        fontFamily = regular,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()

    val buttonText: TextStyle = TextStyle(
        fontSize = 18.sp,
        lineHeight = 22.sp,
        fontFamily = bold,
        fontWeight = FontWeight.W400,
    )

    val inputText: TextStyle = TextStyle(
        fontSize = 16.sp,
        lineHeight = 25.sp,
        fontFamily = regular,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()

    val hintText: TextStyle = TextStyle(
        fontSize = 16.sp,
        lineHeight = 28.sp,
        fontFamily = regular,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()
}

@Immutable
data class MenuTypography(
    private val regular: FontFamily,
    private val bold: FontFamily
) {
    val title: TextStyle = TextStyle(
        fontSize = 24.sp,
        lineHeight = 22.sp,
        fontFamily = regular,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()

    val tableDescription: TextStyle = TextStyle(
        fontSize = 16.sp,
        lineHeight = 22.sp,
        fontFamily = regular,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()

    val itemKey: TextStyle = TextStyle(
        fontSize = 24.sp,
        lineHeight = 22.sp,
        fontFamily = bold,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()

    val itemValue: TextStyle = TextStyle(
        fontSize = 24.sp,
        lineHeight = 22.sp,
        fontFamily = regular,
        fontWeight = FontWeight.W400
    ).preciseLineHeight()
}

fun TextStyle.preciseLineHeight(): TextStyle = this.copy(
    platformStyle = PlatformTextStyle(
        includeFontPadding = false
    ),
    lineHeightStyle = LineHeightStyle(
        alignment = LineHeightStyle.Alignment.Center,
        trim = LineHeightStyle.Trim.None
    )
)